### PR TITLE
[REG-2168] Support a bot action to 'RestartGame'

### PR DESCRIPTION
**Pull request by RG-nAmKcAz**
- Adds support for 'RestartGame' and 'QuitGame' Bot Actions
  - Restart game will restart the game in the editor, or will load a custom impl of what to do in a standalone runtime
  - Quit game will stop the game in the editor, or exit the application in a standalone runtime
- Also, refactors the package membership of criteria and actions to make them easier to find in the explorer .. this should be mostly contained to the first commit if you want to avoid the noise

---

Find the pull request instructions [here](https://www.notion.so/regressiongg/Pull-Request-Process-0d57a7eb582a446983edfacafa406f1e?pvs=4)

Every reviewer and the owner of the PR should consider these points in their request (feel free to copy this checklist so you can fill it out yourself in the overall PR comment)

- [ ] The code is extensible and backward compatible
- [ ] New public interfaces are extensible and open to backward compatibility in the future
- [ ] If preparing to remove a field in the future (i.e. this PR removes an argument), the argument stays but is no longer functional, and attaches a deprecation warning. A linear task is also created to track this deletion task.
- [ ] Non-critical or potentially modifiable arguments are optional
- [ ] Breaking changes and the approach to handling them have been verified with the team (in the Linear task, design doc, or PR itself)
- [ ] The code is easy to read
- [ ] Unit tests are added for expected and edge cases
- [ ] Integration tests are added for expected and edge cases
- [ ] Functions and classes are documented
- [ ] Migrations for both up and down operations are completed
- [ ] A documentation PR is created and being reviewed for anything in this PR that requires knowledge to use
- [ ] Implications on other dependent code (i.e. sample games and sample bots) is considered, mentioned, and properly handled
- [ ] Style changes and other non-blocking changes are marked as non-blocking from reviewers

### CODE REVIEW

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/BotActions/RestartGameBotActionData.cs [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+                        {
+                            var a = allAssemblies[i];
+                            foreach (var type in a.GetTypes())
+                            {
+                                if (typeof(IRGRestartGameAction).IsAssignableFrom(type))
+                                {
+                                    RGDebug.LogInfo($"Using the '{type.FullName}' implementation of IRGRestartGameAction to process Bot Restart Game Actions.  If this is not the type you expected to handle this action, you may accidentally have more than one implementation of the IRGRestartGameAction interface in your runtime.");
+                                    _action = (IRGRestartGameAction)Activator.CreateInstance(type);
+                                }
+                                break;
```
Comment (addisonbgross) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/BotActions/RestartGameBotActionData.cs:58]: do we want this `break` inside the `if` above it?

**Review by addisonbgross — APPROVED** [review]
Looks good overall. Just 1 non-blocking Q

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/BotActions/RestartGameBotActionData.cs (hunk 1 above)
Comment (RG-nAmKcAz) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/BotActions/RestartGameBotActionData.cs:58]: ha.. yeah probably

**Review by vontell — CHANGES_REQUESTED** [review]
Very nice!! I only had one question about default behavior in the editor for the restart mode, but looks good otherwise. Will resolve that quickly once I get an answer.

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/BotActions/RestartGameBotActionData.cs [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+#if UNITY_EDITOR
+
+                RGDebug.LogInfo($"Restarting the game in the editor...");
+                _error = null;
+                error = _error;
+                // register our hook so that the game will start right back up
+                EditorApplication.playModeStateChanged += PlayGameInEditor;
+                // stop the game in the editor
+                EditorApplication.isPlaying = false;
+                return true;
```
[the remark this answers is not in the document]
Comment (RG-nAmKcAz) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/BotActions/RestartGameBotActionData.cs:107]: correct.. as of right now.. 

- Editor - 'restart' actually restarts the game
- External Runtime - 
  - Default - 'restart' loads scene 0, which is as close as we can get to restarting the game without ending the process runtime; this restarts the game to the starting scene, but potentially without cleaning up many things as noted in the comments in the code... we could explore making this much harsher... like.. destroy all objects (event don't destroy on load) before loading scene 0 which would make it much closer to the editor behaviour... however, this still wouldn't kill background threads, or unity task running on entities/etc
  - Custom - 'restart' runs whatever commands are in your custom impl

[the remark this answers is not in the document]
Comment (vontell) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/BotActions/RestartGameBotActionData.cs:107]: I suppose that was fine for now, but I was thinking that even in editor it would run the custom impl, since my restart logic might just be "respawn at the respawn point and remove all enemies". But on second thought, it might be good to allow the custom action in the editor because this means the only way for you to test the custom logic is to keep rebuilding the game, which could take a long time.

[the remark this answers is not in the document]
Comment (RG-nAmKcAz) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/BotActions/RestartGameBotActionData.cs:107]: Apparently we weren't on the same page at all when initially discussing this task.

It seems you wanted an action to define 'reset for next test', but I got 'restart game' out of the initial discussion.

I guess I'll start over .

[the remark this answers is not in the document]
Comment (vontell) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/BotActions/RestartGameBotActionData.cs:107]: Start over? I'm confused as I feel like all the code here aligns with exactly what I was thinking, including the default of loading the initial scene. From what I see here, this aligns almost exactly with my thinking but I had in my mind one small extension, but if that extension is not small I'd rather have this PR in.

**Comment by RG-nAmKcAz** [comment]
@vontell I debated whether to start a new PR for this, but I re-opened the existing.  There is now another new action `ResetForNextTest` which should fulfill what you were looking for, but I am leaving the RestartGame and QuitGame actions in here as they feel safer / more appropriate to me.

You can now do what you intended with a partial reset for the next test without restarting the game, but I don't really agree with this function being different than restart game as there is generally no 1 single action you can take to safely reset the game's state other than fully restarting it.

I feel that this is only going to add confusion for users on how to implement 'reset' safely for multiple test scenarios.

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/BotActionType.cs [hunk 3]
```diff
@@ -11,6 +11,9 @@ public enum BotActionType
         Mouse_CVImage,
         Mouse_CVText,
         Mouse_ObjectDetection,
+        RestartGame,
+        QuitGame,
+        ResetForNextTest
```
Comment (RG-nAmKcAz) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/BotActionType.cs:16]: I would rather name this `ResetGame`, but that is too close to `RestartGame` to be completely clear.

Ultimately, I don't like `Reset` existing separately from `Restart`, but that was the only way to fufill @vontell 's ask that the custom impl be testable in the editor if desired.

**Review by vontell — APPROVED** [review]

**Comment by vontell** [comment]
Approving but happy to discuss this further at some point, I have a feeling what I'm about to write out via text is better explained over a quick huddle at some point.

I think where the confusion came from is what we were assuming would go in this RestartGame custom impl, and I am fine going with whatever you were expecting inside of that... but anyhow, I was expecting the result of my message to not be an entirely new type, it was simply arguing that the custom impl would run in the editor - I guess I am missing some nuance around why we couldn't just delete that UNITY_EDITOR conditional in the ResetGame handler, that was what I thought would be the change (i.e. just making sure if I provide a custom impl, it uses that in editor so I can make sure that works before building). I suppose that means that the sentence "Ultimately, I don't like Reset existing separately from Restart, but that was the only way to fufill @vontell 's ask that the custom impl be testable in the editor if desired" is the part I don't understand, I'll dive a little deeper into the code though and see if I can figure that out for myself
